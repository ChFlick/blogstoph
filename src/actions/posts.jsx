import database from '../firebase/firebase';

export const removePost = (id) => ({
    type: 'REMOVE_POST',
    id
});

export const startRemovePost = (post) => {
    return (dispatch, getState) => {
        const visibility = post.published ? 'public' : 'private';
        return database.ref(`posts/${visibility}/${post.id}`).remove().then((ref) => {
            dispatch(removePost(post.id));
        });
    };
};

export const addPost = (post) => ({
    type: 'ADD_POST',
    post
});

export const startAddPost = (post) => {
    const { title, subtitle, content, author, date, published } = post;

    return (dispatch, getState) => {
        const visibility = published ? 'public' : 'private';
        return database.ref('posts/' + visibility).push({ title, subtitle, content, author, date }).then((ref) => {
            dispatch(addPost({
                id: ref.key,
                published,
                ...post
            }));
        });
    };
};

export const editPost = (id, updates) => ({
    type: 'EDIT_POST',
    id,
    updates
});

export const startEditPost = (id, post) => {
    const { title, subtitle, content, author, date, published } = post;

    return (dispatch, getState) => {
        const prevPostPublished = getState().posts.find((post) => post.id === id).published;
        const visibility = published ? 'public' : 'private';

        if (prevPostPublished !== published) {
            const prevPostVisibility = prevPostPublished ? 'public' : 'private';
            database.ref(`posts/${prevPostVisibility}/${id}`).remove();
        }

        return database.ref(`posts/${visibility}/${id}`).set({ title, subtitle, content, author, date }).then((ref) => {
            dispatch(editPost(
                id,
                post
            ));
        });
    };
};

export const setPosts = (posts) => ({
    type: 'SET_POSTS',
    posts
});

export const startSetPosts = () => {
    const addPostFromSnapshot = (published, posts) => (postSnapshot) => {
        posts.push({
            id: postSnapshot.key,
            published: !!published,
            ...postSnapshot.val()
        });
    };

    return (dispatch, getState) => {
        return Promise.all(getPostsValueDBReferenceByAuth(getState)
        ).then(([publicPostsSnapshot, privatePostsSnapshot]) => {
            const posts = [];
            publicPostsSnapshot.forEach(addPostFromSnapshot(true, posts));

            if (privatePostsSnapshot) {
                privatePostsSnapshot.forEach(addPostFromSnapshot(false, posts));
            }

            dispatch(setPosts(posts));
        });
    };
};

function getPostsValueDBReferenceByAuth(getState) {
    return (getState().auth && getState().auth.uid) ?
        [database.ref('posts/public').once('value'), database.ref('posts/private').once('value')] :
        [database.ref('posts/public').once('value')];
}

