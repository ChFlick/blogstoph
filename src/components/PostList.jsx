import React, { Fragment } from 'react';
import { connect } from 'react-redux';

import PostListItem from './PostListItem';

const isNotLastElement = (index, numberOfPosts) => {
    return index !== numberOfPosts - 1;
};

export const PostList = (props) => (
    <div className="content-container">
        {
            props.posts.length === 0 ? (
                <p>There are no posts available</p>
            ) : (
                props.posts.map((post, index) =>
                    <Fragment key={post.id}>
                        <PostListItem post={post} />
                        {isNotLastElement(index, props.posts.length) && <hr className="bar--light-grey" />}
                    </Fragment>
                )
            )
        }
    </div>
);

const mapStateToProps = (state) => ({
    posts: (state.posts || []).filter((post = {}) => post.published === true)
});

export default connect(mapStateToProps)(PostList);
