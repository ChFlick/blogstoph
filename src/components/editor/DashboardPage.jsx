import React from 'react';

const EditorDashboardPage = () => (
    <div>
        Editor Dasboard page content
    </div>
);

export default EditorDashboardPage;
