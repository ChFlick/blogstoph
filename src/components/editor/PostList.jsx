import React, { Fragment } from 'react';
import { connect } from 'react-redux';

import PostListItem from './PostListItem';

export const PostList = (props) => (
    <Fragment>
        {
            props.posts.length === 0 ? (
                <p>There are no posts available</p>
            ) : (
                props.posts.map((post) => (
                    <PostListItem post={post} key={post.id} />
                ))
            )
        }
    </Fragment>
);

const mapStateToProps = (state) => ({
    posts: state.posts || []
});

export default connect(mapStateToProps)(PostList);
