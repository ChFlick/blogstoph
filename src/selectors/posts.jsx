const getVisiblePosts = (posts = [], { published, sortBy = 'date' } = {}) => {
    return posts.filter((post) => {
        const publishedMatch = published === undefined || (!!published === post.published);

        return publishedMatch;
    }).sort((a, b) => {
        return b[sortBy] - a[sortBy];
    });
};

const getPostByIdOrTitle = (posts, searchField) => {
    return posts.find((post) =>
        post.id === searchField ||
        post.title.toLowerCase().replace(/ /g, '-') === searchField.toLowerCase());
};

export { getVisiblePosts, getPostByIdOrTitle };
