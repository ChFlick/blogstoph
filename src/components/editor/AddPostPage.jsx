import React, { Fragment } from 'react';
import { connect } from 'react-redux';

import { startAddPost } from '../../actions/posts';
import PostForm from './PostForm';

export class AddPostPage extends React.Component {
    goBack = () => {
        this.props.history.push('/editor/dashboard');
    }

    onSubmit = (post) => {
        this.props.addPost(post);
        this.goBack();
    };

    render() {
        return (
            <Fragment>
                <h1>Add Post</h1>
                <PostForm onSubmit={this.onSubmit} onBack={this.goBack} />
            </Fragment>
        );
    };
};

const mapDispatchToProps = (dispatch) => ({
    addPost: (post) => dispatch(startAddPost(post))
});

export default connect(undefined, mapDispatchToProps)(AddPostPage);
