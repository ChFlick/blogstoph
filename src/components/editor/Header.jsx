import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../../actions/auth';


export const EditorHeader = ({ startLogout }) => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <Link className="header__title" to="/editor/dashboard">
                    <h1>Blogstoph</h1>
                </Link>
                <button className="button button--link" onClick={startLogout}>Logout</button>
            </div>
        </div>
    </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(EditorHeader);
