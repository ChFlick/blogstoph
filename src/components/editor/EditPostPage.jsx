import React, { Fragment } from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';

import { startEditPost, startRemovePost } from '../../actions/posts';
import PostForm from './PostForm';
import { getPostByIdOrTitle } from '../../selectors/posts';

export class EditPostPage extends React.Component {
    state = {
        modalIsOpen: false
    };

    openModal = () => {
        this.setState(() => ({ modalIsOpen: true }));
    }

    closeModal = () => {
        this.setState(() => ({ modalIsOpen: false }));
    }

    goBack = () => {
        this.props.history.push('/editor/dashboard');
    };

    onSubmit = (post) => {
        this.props.editPost(this.props.post.id, post);
        this.goBack();
    };

    onRemove = () => {
        this.props.removePost(this.props.post);
        this.goBack();
    };

    pageHeader = (
        <header className="page-header">
            <div className="content-container">
                <h1 className="page-header__title ">Edit Post</h1>
            </div>
        </header>
    );

    createPageContent = () => (
        <section className="content-container">
            <PostForm
                post={this.props.post}
                onSubmit={this.onSubmit}
                onBack={this.goBack} />
            <button
                className="button"
                type="button"
                onClick={this.openModal}
            >
                Remove Post
            </button>
        </section>
    );

    createModal = () => (
        <Modal
            isOpen={this.state.modalIsOpen}
            onRequestClose={this.closeModal}
            contentLabel="Remove Post"
        >
            <section>
                <h2>Remove Post</h2>
                <p>Are you sure you want to remove the post?</p>
                <button onClick={this.closeModal}>Cancel</button>
                <button onClick={this.onRemove}>Ok</button>
            </section>
        </Modal>
    );

    render() {
        return (
            <Fragment>
                {this.pageHeader}
                {this.createPageContent()}
                {this.createModal()}
            </Fragment>
        );
    };
};

const mapStateToProps = (state, props) => ({
    post: getPostByIdOrTitle(state.posts, props.match.params.id)
});

const mapDispatchToProps = (dispatch) => ({
    editPost: (id, post) => dispatch(startEditPost(id, post)),
    removePost: (id) => dispatch(startRemovePost(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPostPage);
