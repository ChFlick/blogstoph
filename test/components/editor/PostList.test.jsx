import React from 'react';
import { shallow } from 'enzyme';

import { publishedPosts } from '../../fixtures/posts';
import { PostList } from '../../../src/components/editor/PostList';

describe('post list', () => {
    test('should render correctly without posts', () => {
        const wrapper = shallow(<PostList posts={[]} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('should render correctly with posts', () => {
        const wrapper = shallow(<PostList posts={publishedPosts} />);

        expect(wrapper).toMatchSnapshot();
    });
});
