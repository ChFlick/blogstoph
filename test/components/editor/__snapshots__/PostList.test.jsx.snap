// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post list should render correctly with posts 1`] = `
<React.Fragment>
  <div
    className="list-header"
  >
    <div>
      Blog posts
    </div>
  </div>
  <div
    className="list-body"
  >
    <PostListItem
      key="1"
      post={
        Object {
          "author": "Author of 1",
          "content": "This is a Content of a Post...",
          "date": 0,
          "id": 1,
          "published": true,
          "subtitle": "A great subtitle",
          "title": "Some Title",
        }
      }
    />
    <PostListItem
      key="2"
      post={
        Object {
          "author": "Author of 2",
          "content": "This is a second Content of a Post...",
          "date": 1000,
          "id": 2,
          "published": true,
          "subtitle": "A great subtitle",
          "title": "Some Title 2",
        }
      }
    />
    <PostListItem
      key="3"
      post={
        Object {
          "author": "Author of 3",
          "content": "This is a third Content of a Post...",
          "date": 2000,
          "id": 3,
          "published": true,
          "subtitle": "A great subtitle",
          "title": "Some Title 3",
        }
      }
    />
  </div>
</React.Fragment>
`;

exports[`post list should render correctly without posts 1`] = `
<React.Fragment>
  <div
    className="list-header"
  >
    <div>
      Blog posts
    </div>
  </div>
  <div
    className="list-body"
  >
    <p
      className="list-item list-item--message"
    >
      There are no posts available
    </p>
  </div>
</React.Fragment>
`;
